### DISCUSSION TITLE
Converter is not being called on Binding

### DISCUSSION THREAD

**Discussion by ArchLeaders** [AvaloniaUI/Avalonia#9316] [Q&A] [answered]
Hello,

I recently attempted to use a converter in Avalonia to load a dynamic image, but the converter is never called (and I suspect not loading/registering at all).

Here's how it's setup:

I have the Converter class here: `BotwSaveManager/Converters/BitmapAssetValueConverter.cs`
```cs
/// <summary>
/// <para>
/// Converts a string path to a bitmap asset.
/// </para>
/// <para>
/// The asset must be in the same assembly as the program. If it isn't,
/// specify "avares://<assemblynamehere>/" in front of the path to the asset.
/// </para>
/// </summary>
public class BitmapAssetValueConverter : IValueConverter
{
    public static BitmapAssetValueConverter Instance = new();

    public object? Convert(object? value, Type? targetType, object? parameter, CultureInfo culture)
    {
        if (value == null)
            return null;

        if (value is string rawUri && targetType == typeof(IBitmap)) {
            Uri uri;

            // Allow for assembly overrides
            if (rawUri.StartsWith("avares://")) {
                uri = new Uri(rawUri);
            }
            else {
                string? assemblyName = Assembly.GetEntryAssembly()?.GetName().Name;
                uri = new Uri($"avares://{assemblyName}{rawUri}");
            }

            var assets = AvaloniaLocator.Current.GetService<IAssetLoader>();
            if (assets != null) {
                var asset = assets.Open(uri);
                return new Bitmap(asset);
            }
        }

        throw new NotSupportedException();
    }

    public object ConvertBack(object? value, Type? targetType, object? parameter, CultureInfo culture)
    {
        throw new NotSupportedException();
    }
}
```

And then in my AXAML file: `BotwSaveManager/Views/BotwSaveView.axaml`
```xml
<UserControl x:Class="BotwSaveManager.Views.BotwSaveView"
             xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:ext="clr-namespace:BotwSaveManager.Converters;assembly=BotwSaveManager"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             d:DesignHeight="450"
             d:DesignWidth="800"
             mc:Ignorable="d">
    <UserControl.Resources>
        <ext:BitmapAssetValueConverter x:Key="VariableImage" /> <!-- Create an instance of the converter? -->
    </UserControl.Resources>
    <Grid>
        <Image Width="75"
               Height="73"
               Source="{Binding Image, Converter={StaticResource VariableImage}}" /> <!-- Use it? -->
    </Grid>
</UserControl>
```

And finally I have the ViewModel: `BotwSaveManager/Views/BotwSaveViewModel.cs` (triple-checked, it is assigned to the view in the code behind)
```cs
public class BotwSaveViewModel : ReactiveObject
{
    public BotwSave BotwSave { get; set; }

    private string image;
    public string Image {
        get => image;
        set => this.RaiseAndSetIfChanged(ref image, value);
    }

    public BotwSaveViewModel(BotwSave save)
    {
        BotwSave = save;
        Image = $"avares://Assets/{save.SaveType}.png";
    }
}
```

As far as I can tell, this is how to use Converters, but I cannot get it to work at all. The converter code is never run (checked with a stop) and there are no related compiler errors/warnings, it compiles and runs fine except the referenced image never shows up.
Also, the referenced image _does_ exists, I have checked the spelling and properties (AvaloniaResource).

![image](https://user-images.githubusercontent.com/80713508/199208168-92aabe6c-ebd6-4d75-a456-cc1ac2f20f1d.png)

Supposedly I've missed something really obvious here, but after two hours I couldn't find it, so a point in the right direction would be greatly appreciated.

Thanks.

**Comment by timunie** [comment, marked as answer]
Can you add a TextBlock bimding to Image so we can see if the DataContrxt is there at all?

**Reply by ArchLeaders** [reply]
> Can you add a TextBlock bimding to Image so we can see if the DataContrxt is there at all?

Yeah okay, I'm just an idiot, went and did that with no luck, I then decided to just leave the TextBox with static text, and it didn't show up either. Went to investigate, and I found that I forgot to copy `InitializeComponent();` into my custom view constructor.

Guess that's what I get for doing stuff at 3 AM. Thanks for bringing my attention to that one!

---

On a related note, although I haven't tried to fix this one yet; when the Converter runs, the `targetType` comes in as an `Avalonia.Media.IImage`, anyway I can get this to a `Bitmap`?

**Reply by timunie** [reply]
I think Bitmap implements IImage, so it should be safe to cast.
